source/common/filter/config_discovery_impl.h

```diff
@@ -233,17 +233,35 @@ class StaticFilterConfigProviderImpl : public FilterConfigProvider {
   const std::string filter_config_name_;
 };
 
+/**
+ * Base class for a FilterConfigProviderManager.
+ */
+class FilterConfigProviderManagerImplBase : Logger::Loggable<Logger::Id::filter> {
+protected:
+  std::shared_ptr<FilterConfigSubscription>
+  getSubscription(const envoy::config::core::v3::ConfigSource& config_source,
+                  const std::string& name, Server::Configuration::FactoryContext& factory_context,
+                  const std::string& stat_prefix);
+  void applyLastOrDefaultConfig(std::shared_ptr<FilterConfigSubscription>& subscription,
+                                DynamicFilterConfigProviderImplBase& provider,
+                                const std::string& filter_config_name);
+
+private:
+  absl::flat_hash_map<std::string, std::weak_ptr<FilterConfigSubscription>> subscriptions_;
+  friend class FilterConfigSubscription;
+};
+
 /**
  * An implementation of FilterConfigProviderManager.
  */
-class FilterConfigProviderManagerImpl : public FilterConfigProviderManager,
-                                        public Singleton::Instance,
-                                        Logger::Loggable<Logger::Id::filter> {
+class FilterConfigProviderManagerImpl : public FilterConfigProviderManagerImplBase,
+                                        public FilterConfigProviderManager,
+                                        public Singleton::Instance {
 public:
   DynamicFilterConfigProviderPtr createDynamicFilterConfigProvider(
       const envoy::config::core::v3::ExtensionConfigSource& config_source,
       const std::string& filter_config_name, Server::Configuration::FactoryContext& factory_context,
-      const std::string& stat_prefix, bool last_filter_in_filter_chain,
+      const std::string& stat_prefix, bool last_filter_in_filter_config,
```

Comment:
I think we should keep the name `last_filter_in_filter_chain`. The condition is that only terminal filters are last in chains.

Reply:
Ah ok. I did it due to [this](https://github.com/envoyproxy/envoy/blob/7796988e02c048982f3c8cda62067db1a113c9fb/source/common/filter/config_discovery_impl.cc#L252) since that code is going to be moved into the header. When I do that I can rename config -> chain from the code's body and elsewhere in the file as well

Reply:
So the plan is to handle this in a follow up or in this PR?

Reply:
A follow up. `config` is used in multiple other places too, not just here

Reply:
SGTM, let's merge this one then